File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/FavoritePlayersManagerImpl.java
package com.garpr.android.misc;

import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;

import com.garpr.android.R;
import com.garpr.android.models.AbsPlayer;
import com.garpr.android.preferences.KeyValueStore;
import com.google.gson.Gson;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class FavoritePlayersManagerImpl implements FavoritePlayersManager {

    private static final String TAG = "FavoritePlayersManagerImpl";

    private final Gson mGson;
    private final KeyValueStore mKeyValueStore;
    private final List<WeakReference<OnFavoritePlayersChangeListener>> mListeners;
    private final Timber mTimber;


    public FavoritePlayersManagerImpl(@NonNull final Gson gson,
            @NonNull final KeyValueStore keyValueStore, @NonNull final Timber timber) {
        mGson = gson;
        mKeyValueStore = keyValueStore;
        mListeners = new LinkedList<>();
        mTimber = timber;
    }

    @Override
    public void addListener(@NonNull final OnFavoritePlayersChangeListener listener) {
        synchronized (mListeners) {
            boolean addListener = true;
            final Iterator<WeakReference<OnFavoritePlayersChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnFavoritePlayersChangeListener> reference = iterator.next();
                final OnFavoritePlayersChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else if (item == listener) {
                    addListener = false;
                }
            }

            if (addListener) {
                mListeners.add(new WeakReference<>(listener));
            }
        }
    }

    @Override
    public void addPlayer(@NonNull final AbsPlayer player) {
        if (containsPlayer(player)) {
            mTimber.d(TAG, "Not adding favorite, it already exists in the store");
            return;
        }

        final List<AbsPlayer> players = getPlayers();
        mTimber.d(TAG, "Adding favorite (there are currently " +
                (players == null ? 0 : players.size()) + " favorite(s))");

        final String playerJson = mGson.toJson(player, AbsPlayer.class);
        mKeyValueStore.setString(player.getId(), playerJson);
        notifyListeners();
    }

    @Override
    public void clear() {
        mKeyValueStore.clear();
        notifyListeners();
    }

    @Override
    public boolean containsPlayer(@NonNull final AbsPlayer player) {
        return containsPlayer(player.getId());
    }

    @Override
    public boolean containsPlayer(@NonNull final String playerId) {
        return mKeyValueStore.contains(playerId);
    }

    @Nullable
    @Override
    public List<AbsPlayer> getPlayers() {
        final Map<String, ?> all = mKeyValueStore.getAll();

        if (all == null || all.isEmpty()) {
            return null;
        }

        final List<AbsPlayer> players = new ArrayList<>(all.size());

        for (final Map.Entry<String, ?> entry : all.entrySet()) {
            final String json = (String) entry.getValue();
            players.add(mGson.fromJson(json, AbsPlayer.class));
        }

        Collections.sort(players, AbsPlayer.ALPHABETICAL_ORDER);
        return players;
    }

    @Override
    public boolean isEmpty() {
        final Map<String, ?> all = mKeyValueStore.getAll();
        return all == null || all.isEmpty();
    }

    private void notifyListeners() {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnFavoritePlayersChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnFavoritePlayersChangeListener> reference = iterator.next();
                final OnFavoritePlayersChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else {
                    item.onFavoritePlayersChanged(this);
                }
            }
        }
    }

    @Override
    public void removeListener(@Nullable final OnFavoritePlayersChangeListener listener) {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnFavoritePlayersChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnFavoritePlayersChangeListener> reference = iterator.next();
                final OnFavoritePlayersChangeListener item = reference.get();

                if (item == null || item == listener) {
                    iterator.remove();
                }
            }
        }
    }

    @Override
    public void removePlayer(@NonNull final AbsPlayer player) {
        removePlayer(player.getId());
    }

    @Override
    public void removePlayer(@NonNull final String playerId) {
        mKeyValueStore.remove(playerId);
        notifyListeners();
    }

    @Override
    public boolean showAddOrRemovePlayerDialog(@NonNull Context context,
            @Nullable final AbsPlayer player) {
        if (player == null) {
            return false;
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(context)
                .setNegativeButton(R.string.cancel, null);

        if (containsPlayer(player)) {
            builder.setMessage(context.getString(R.string.remove_x_from_favorites, player.getName()))
                    .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(final DialogInterface dialog, final int which) {
                            removePlayer(player);
                        }
                    });
        } else {
            builder.setMessage(context.getString(R.string.add_x_to_favorites, player.getName()))
                    .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(final DialogInterface dialog, final int which) {
                            addPlayer(player);
                        }
                    });
        }

        builder.show();
        return true;
    }

    @Override
    public int size() {
        final Map<String, ?> all = mKeyValueStore.getAll();
        return all == null ? 0 : all.size();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/FavoritePlayersManagerImpl.kt
package com.garpr.android.misc

import android.content.Context
import android.support.v7.app.AlertDialog
import com.garpr.android.R
import com.garpr.android.misc.FavoritePlayersManager.OnFavoritePlayersChangeListener
import com.garpr.android.models.AbsPlayer
import com.garpr.android.models.FavoritePlayer
import com.garpr.android.models.Region
import com.garpr.android.preferences.KeyValueStore
import com.google.gson.Gson
import java.lang.ref.WeakReference
import java.util.*

class FavoritePlayersManagerImpl(
        private val gson: Gson,
        private val keyValueStore: KeyValueStore,
        private val timber: Timber
) : FavoritePlayersManager {

    private val listeners = mutableListOf<WeakReference<OnFavoritePlayersChangeListener>>()


    companion object {
        private const val TAG = "FavoritePlayersManagerImpl"
    }

    override val absPlayers: List<AbsPlayer>?
        get() {
            val players = this.players

            if (players == null || players.isEmpty()) {
                return null
            }

            val list = mutableListOf<AbsPlayer>()
            list.addAll(players)

            return list
        }

    override fun addListener(listener: OnFavoritePlayersChangeListener) {
        synchronized (listeners) {
            var addListener = true
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else if (item == listener) {
                    addListener = false
                }
            }

            if (addListener) {
                listeners.add(WeakReference(listener))
            }
        }
    }

    override fun addPlayer(player: AbsPlayer, region: Region) {
        if (player in this) {
            timber.d(TAG, "Not adding favorite, it already exists in the store")
            return
        }

        timber.d(TAG, "Adding favorite (there are currently $size favorite(s))")

        val favoritePlayer = FavoritePlayer(player.id, player.name, region)
        val playerJson = gson.toJson(favoritePlayer, FavoritePlayer::class.java)
        keyValueStore.setString(player.id, playerJson)
        notifyListeners()
    }

    override fun clear() {
        keyValueStore.clear()
        notifyListeners()
    }

    override fun contains(player: AbsPlayer): Boolean {
        return player.id in this
    }

    override fun contains(playerId: String): Boolean {
        return playerId in keyValueStore
    }

    override val isEmpty: Boolean
        get() {
            val all = keyValueStore.all
            return all == null || all.isEmpty()
        }

    private fun notifyListeners() {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else {
                    item.onFavoritePlayersChanged(this)
                }
            }
        }
    }

    override val players: List<FavoritePlayer>?
        get() {
            val all = keyValueStore.all

            if (all == null || all.isEmpty()) {
                return null
            }

            val players = mutableListOf<FavoritePlayer>()

            for ((key, value) in all) {
                val json = value as String
                players.add(gson.fromJson(json, FavoritePlayer::class.java))
            }

            Collections.sort(players, AbsPlayer.ALPHABETICAL_ORDER)
            return players
        }

    override fun removeListener(listener: OnFavoritePlayersChangeListener?) {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null || item == listener) {
                    iterator.remove()
                }
            }
        }
    }

    override fun removePlayer(player: AbsPlayer) {
        removePlayer(player.id)
    }

    override fun removePlayer(playerId: String) {
        keyValueStore.remove(playerId)
        notifyListeners()
    }

    override fun showAddOrRemovePlayerDialog(context: Context, player: AbsPlayer?, region: Region): Boolean {
        if (player == null) {
            return false
        }

        val builder = AlertDialog.Builder(context)
                .setNegativeButton(R.string.cancel, null)

        if (player in this) {
            builder.setMessage(context.getString(R.string.remove_x_from_favorites, player.name))
                    .setPositiveButton(R.string.yes) { dialog, which -> removePlayer(player) }
        } else {
            builder.setMessage(context.getString(R.string.add_x_to_favorites, player.name))
                    .setPositiveButton(R.string.yes) { dialog, which -> addPlayer(player, region) }
        }

        builder.show()
        return true
    }

    override val size: Int
        get() = keyValueStore.all?.size ?: 0

}
